test/e2e/stack_test.go

```diff
@@ -0,0 +1,219 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package e2e
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	apmv1 "github.com/elastic/cloud-on-k8s/pkg/apis/apm/v1"
+	beatv1beta1 "github.com/elastic/cloud-on-k8s/pkg/apis/beat/v1beta1"
+	commonv1 "github.com/elastic/cloud-on-k8s/pkg/apis/common/v1"
+	esv1 "github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1"
+	entv1beta1 "github.com/elastic/cloud-on-k8s/pkg/apis/enterprisesearch/v1beta1"
+	kbv1 "github.com/elastic/cloud-on-k8s/pkg/apis/kibana/v1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/beat/filebeat"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/version"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+	beattests "github.com/elastic/cloud-on-k8s/test/e2e/beat"
+	"github.com/elastic/cloud-on-k8s/test/e2e/test"
+	"github.com/elastic/cloud-on-k8s/test/e2e/test/apmserver"
+	"github.com/elastic/cloud-on-k8s/test/e2e/test/beat"
+	"github.com/elastic/cloud-on-k8s/test/e2e/test/elasticsearch"
+	"github.com/elastic/cloud-on-k8s/test/e2e/test/enterprisesearch"
+	"github.com/elastic/cloud-on-k8s/test/e2e/test/kibana"
+)
+
+const (
+	// upgrading the entire stack can take some time, since we need to account for (in order):
+	// - Elasticsearch rolling upgrade
+	// - Kibana + Enterprise Search deployments upgrade
+	// - APMServer deployment upgrade + Beat daemonset upgrade
+	stackVersionUpgradeTimeout = 30 * time.Minute
+)
+
+// TestVersionUpgradeOrdering deploys the entire stack, with resources associated together.
+// Then, it updates their version, and ensures a strict ordering is respected during the version upgrade.
+func TestVersionUpgradeOrdering(t *testing.T) {
+	initialVersion := "7.7.0"
+	updatedVersion := "7.8.0"
+
+	// Single-node ES clusters cannot be green with APM indices (see https://github.com/elastic/apm-server/issues/414).
+	es := elasticsearch.NewBuilder("es").
+		WithESMasterDataNodes(3, elasticsearch.DefaultResources).
+		WithVersion(initialVersion)
+	esUpdated := es.WithVersion(updatedVersion)
+	esRef := commonv1.ObjectSelector{Namespace: es.Elasticsearch.Namespace, Name: es.Elasticsearch.Name}
+	kb := kibana.NewBuilder("kb").
+		WithNodeCount(1).
+		WithVersion(initialVersion).
+		WithElasticsearchRef(esRef)
+	kbUpdated := kb.WithVersion(updatedVersion)
+	kbRef := commonv1.ObjectSelector{Namespace: kb.Kibana.Namespace, Name: kb.Kibana.Name}
+	apm := apmserver.NewBuilder("apm").
+		WithNodeCount(1).
+		WithVersion(initialVersion).
+		WithElasticsearchRef(esRef).
+		WithKibanaRef(kbRef)
+	apmUpdated := apm.WithVersion(updatedVersion)
+	ent := enterprisesearch.NewBuilder("ent").
+		WithNodeCount(1).
+		WithVersion(initialVersion).
+		WithElasticsearchRef(esRef)
```

Comment:
`WithRestrictedSecurityContext` might be needed in a few places:

```suggestion
		WithElasticsearchRef(esRef).
		WithRestrictedSecurityContext()
```